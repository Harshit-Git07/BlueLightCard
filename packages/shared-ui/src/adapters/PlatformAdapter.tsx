import { createContext, useContext } from 'react';
import { AmplitudeLogParams, PlatformVariant } from '../types';

export type HttpMethod = 'GET' | 'PUT' | 'POST' | 'DELETE' | 'PATCH' | 'HEAD' | 'OPTIONS' | 'TRACE';

export type V5RequestOptions = {
  /**
   * Platforms should not cache responses if the cachePolicy is set to 'never'.
   * Setting the cachePolicy to 'auto' allows the platform to cache responses but
   * does not require it to do so. If this option is omitted, platforms should
   * default to 'never'.
   *
   * @default 'never'
   */
  cachePolicy?: 'never' | 'auto';
  method: HttpMethod;
  queryParameters?: Record<string, string>;
  body?: string;
};

export type V5Response = {
  statusCode: number;
  body: string;
};

export interface IPlatformAdapter {
  /**
   * Can be used by the shared-ui to introspect which platform it is running in
   */
  platform: PlatformVariant;
  /**
   * Invokes a v5 API endpoint
   */
  invokeV5Api(path: string, options: V5RequestOptions): Promise<V5Response>;
  /**
   * Logs analytics events
   */
  logAnalyticsEvent(event: string, parameters: AmplitudeLogParams): void;
  /**
   * Navigate to a route within the app
   */
  navigate(path: string): void;
  /**
   * Navigate to an external URL
   */
  navigateExternal(url: string): void;
}

const PlatfromAdapterContext = createContext<IPlatformAdapter | null>(null);

type Props = {
  children: React.ReactNode;
  adapter: IPlatformAdapter;
};

export function PlatformAdapterProvider({ adapter, children }: Props): React.ReactNode {
  return (
    <PlatfromAdapterContext.Provider value={adapter}>{children}</PlatfromAdapterContext.Provider>
  );
}

export function usePlatformAdapter() {
  const adapter = useContext(PlatfromAdapterContext);

  if (!adapter) {
    throw new Error(
      'Missing platform adapter implementation. Did you forget to wrap your app with <PlatformAdapterProvider>?',
    );
  }

  return adapter;
}
