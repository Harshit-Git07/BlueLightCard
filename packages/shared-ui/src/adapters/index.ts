export * from './PlatformAdapter';
export * from './useMockPlatformAdapter';
export * from './StorybookPlatformAdapter';
