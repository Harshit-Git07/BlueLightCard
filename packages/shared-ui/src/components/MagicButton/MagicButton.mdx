import { Meta, Canvas, Controls } from '@storybook/blocks';

import * as MagicButton from './MagicButton.stories';

<Meta of={MagicButton} />

# MagicButton

- [Overview](#overview)
- [Props](#props)
- [Usage](#usage)

## Overview
Overview description of the MagicButton component...

## Props
<Controls />

## Usage
➡️ Usage instruction one

➡️ Usage instruction two

➡️ Usage instruction three

➡️ Usage instruction four

## Variants
List of variants can be shown below here.

### Primary
<Canvas of={MagicButton.Primary} />

### Pressed
<Canvas of={MagicButton.Pressed} />

### Disabled
<Canvas of={MagicButton.Disabled} />

