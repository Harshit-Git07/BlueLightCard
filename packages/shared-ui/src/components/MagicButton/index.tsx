import { FC, PropsWithChildren } from 'react';
import { IconDefinition } from '@fortawesome/pro-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

export enum MagicBtnVariant {
  Primary = 'primary',
  Pressed = 'pressed',
  Disabled = 'disabled',
}

type BodyStyles = {
  [key in MagicBtnVariant]: string;
};

export type Props = PropsWithChildren & {
  variant: MagicBtnVariant;
  onClick?: () => void;
  clickable?: boolean;
  className?: string;
  label: string;
  description?: string;
  icon?: IconDefinition;
};

const MagicButton: FC<Props> = ({
  variant = MagicBtnVariant.Primary,
  clickable = true,
  onClick = undefined,
  className = '',
  icon,
  label,
  description,
}) => {
  const pressedBtn = variant === MagicBtnVariant.Pressed;
  const disabledBtn = variant === MagicBtnVariant.Disabled;

  // **** BUTTON COLOUR TOKENS ***
  // ---- Labels for primary and pressed variants ----
  const primaryBtnLabelColour =
    'text-magicButton-default-label-colour dark:text-magicButton-default-label-colour-dark';
  const pressedBtnLabelColour =
    'text-magicButton-pressed-label-colour dark:text-magicButton-pressed-label-colour-dark';
  // ---- Description for pressed variant ----
  const pressedBtnDescriptionColour =
    'text-magicButton-pressed-description-colour dark:text-magicButton-pressed-description-colour-dark';
  // ****

  // **** BUTTON TYPOGRAPHY TOKENS FOR LABEL AND DESCRIPTION ***
  const btnLabelTypography =
    'font-magicButton-label-font font-magicButton-label-font-weight text-magicButton-label-font tracking-magicButton-label-font leading-magicButton-label-font';
  const btnDescriptionTypography =
    'font-magicButton-description-font font-magicButton-description-font-weight text-magicButton-description-font tracking-magicButton-description-font leading-magicButton-description-font';
  // ****

  // **** BUTTON GRADIENT TOKENS ***
  const gradientBgPrimaryBtn =
    'bg-gradient-to-b from-colour-primary-gradient-bright-fixed dark:from-colour-primary-gradient-bright-fixed to-colour-primary-gradient-dim-fixed dark:to-colour-primary-gradient-dim-fixed';
  const gradientAnimatedPressedBtn =
    'bg-gradient-to-r from-5% to-95% from-colour-secondary-gradient-bright-fixed dark:from-colour-secondary-gradient-bright-fixed via-colour-secondary-gradient-centre-fixed dark:via-colour-secondary-gradient-centre-fixed to-colour-secondary-gradient-dim-fixed dark:to-colour-secondary-gradient-dim-fixed';
  // ****

  const bodyStyles: BodyStyles = {
    primary: `${gradientBgPrimaryBtn} hover:opacity-90 transition-opacity duration-200`,
    pressed: 'bg-colour-surface dark:bg-colour-surface-dark',
    disabled:
      'text-magicButton-disabled-label-colour dark:text-magicButton-disabled-label-colour-dark bg-magicButton-disabled-bg-colour dark:bg-magicButton-disabled-bg-colour-dark opacity-30',
  };

  return (
    <button
      className={`${className} z-10 relative w-fit rounded-full overflow-hidden p-1 h-[3.75rem] ${
        disabledBtn ? 'disabled' : clickable ? 'cursor-pointer' : 'cursor-default'
      }`}
      disabled={disabledBtn}
      onClick={() => (!disabledBtn && onClick ? onClick() : null)}
    >
      {!disabledBtn && pressedBtn && (
        <div className="absolute z-0 aspect-square min-w-[125%] min-h-[125%] top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 flex">
          <div
            className={`${gradientAnimatedPressedBtn} z-0 h-full w-full animate-magicButtonGradient`}
            style={{ backgroundSize: '400% 400%' }}
          />
        </div>
      )}
      <div
        className={`relative py-2 px-4 rounded-full z-10 h-full w-full flex justify-center items-center ${bodyStyles[variant]}`}
      >
        {!pressedBtn ? (
          <span className={`${primaryBtnLabelColour} ${btnLabelTypography}`}>{label}</span>
        ) : (
          <div className="flex-col w-full text-nowrap whitespace-nowrap flex-nowrap justify-center items-center">
            <div
              className={`${pressedBtnLabelColour} ${btnLabelTypography} text-center flex justify-center gap-2 items-center`}
            >
              {icon && <FontAwesomeIcon icon={icon} />}
              {label}
            </div>
            <div className={`${pressedBtnDescriptionColour} ${btnDescriptionTypography}`}>
              {description}
            </div>
          </div>
        )}
      </div>
    </button>
  );
};

export default MagicButton;
