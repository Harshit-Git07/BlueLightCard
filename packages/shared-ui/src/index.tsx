export { default as Button } from './components/Button';
export { default as Search } from './components/Search';
export { default as SearchBar } from './components/SearchBar';
export { default as Iso } from './components/Iso';
export { default as Container } from './components/Container';
export { default as PillButtons } from './components/PillButtons';
export { default as CompanyAbout } from './components/CompanyAbout';
export { default as Dropdown } from './components/Dropdown';
export { default as Heading } from './components/Heading';
export { default as Image } from './components/Image';
export { default as ResponsiveOfferCard } from './components/ResponsiveOfferCard';
export { default as Badge } from './components/Badge';
export { default as CampaignCard } from './components/CampaignCard';
export { default as ShareButton } from './components/ShareButton';
export { default as Link } from './components/Link';
export { default as OfferCardList } from './components/OfferCardList';
export { default as OfferSheet } from './components/OfferSheet';
export { default as MagicButton } from './components/MagicButton';
export { default as Label } from './components/Label';
export { default as SwiperCarousel } from './components/SwiperCarousel';
export { default as PillGroup } from './components/PillGroup';
export { default as Typography } from './components/Typography';
export { default as PaginationControls } from './components/PaginationControls';
export { default as EventBus } from './lib/eventBus';

export * from './adapters';
export * from './api';
export * from './components/OfferDetails';
export * from './hooks/useCSS';
export * from './hooks/useMediaQuery';
export * from './utils/cssUtil';
export * from './utils/storyUtils';
export * from './utils/getCDNUrl';
export * from './utils/getPathFromUrl';
export * from './utils/ageUtils';
export * from './lib/rewriters';
export * from './lib/invoke';
export * from './types';
export * from './utils/offers/offerSheetParser';
export * from './context/OfferSheet/OfferSheetContext';
export * from './context/OfferSheet/hooks';
export * from './providers';
export * from './test';
export * from './env';
export * from './utils/darkRead/darkRead';
