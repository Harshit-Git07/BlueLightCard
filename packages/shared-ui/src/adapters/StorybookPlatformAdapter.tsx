import { Decorator } from '@storybook/react';
import { IPlatformAdapter, PlatformAdapterProvider } from './PlatformAdapter';
import { PlatformVariant } from '../types';

export const storybookPlatformAdapter: IPlatformAdapter = {
  getAmplitudeFeatureFlag: () => 'control',
  invokeV5Api: () => Promise.resolve({ status: 200, data: '{}' }),
  logAnalyticsEvent: () => {},
  navigate: () => {},
  navigateExternal: () => ({
    isOpen: () => true,
  }),
  writeTextToClipboard: () => Promise.resolve(),
  getBrandURL: () => '',
  platform: PlatformVariant.MobileHybrid,
} satisfies IPlatformAdapter;

export const StorybookPlatformAdapterDecorator: Decorator = (Story, { parameters }) => {
  const platformAdapter = {
    ...storybookPlatformAdapter,
    ...parameters.platformAdapter,
  };

  return (
    <PlatformAdapterProvider adapter={platformAdapter}>
      <Story platformAdapter={platformAdapter} />
    </PlatformAdapterProvider>
  );
};
