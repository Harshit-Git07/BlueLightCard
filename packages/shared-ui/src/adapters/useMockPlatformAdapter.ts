import { IPlatformAdapter } from './PlatformAdapter';
import { PlatformVariant } from '../types';

export const useMockPlatformAdapter = (
  status = 200,
  data = {},
  platform = PlatformVariant.MobileHybrid,
) => {
  const getAmplitudeFeatureFlag = jest.fn();
  const invokeV5Api = jest.fn().mockResolvedValue({ status, data: JSON.stringify(data) });
  const logAnalyticsEvent = jest.fn();
  const navigate = jest.fn();
  const navigateExternal = jest.fn();
  const writeTextToClipboard = jest.fn().mockReturnValue(Promise.resolve());

  const mockPlatformAdapter = {
    platform,
    getAmplitudeFeatureFlag,
    invokeV5Api,
    logAnalyticsEvent,
    navigate,
    navigateExternal,
    writeTextToClipboard,
  } satisfies IPlatformAdapter;

  return mockPlatformAdapter;
};
