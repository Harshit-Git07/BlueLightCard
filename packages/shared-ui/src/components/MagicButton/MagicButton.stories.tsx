import { Meta, StoryFn } from '@storybook/react';
import MagicButton, { MagicBtnVariant } from './';
import { faWandMagicSparkles } from '@fortawesome/pro-solid-svg-icons';

const componentMeta: Meta<typeof MagicButton> = {
  title: 'Component System/Magic Button',
  component: MagicButton,
  argTypes: {
    variant: {
      description: 'The button variant (Primary, Pressed or Disabled). Primary is the default.',
      options: [MagicBtnVariant.Primary, MagicBtnVariant.Pressed, MagicBtnVariant.Disabled],
    },
    clickable: {
      description: 'Whether the button is clickable. Overriden when disabled.',
    },
    label: {
      description: 'Button label text',
    },
  },
};

const DefaultTemplate: StoryFn<typeof MagicButton> = (args) => <MagicButton {...args} />;

export const Primary = DefaultTemplate.bind({});

Primary.args = {
  variant: MagicBtnVariant.Primary,
  label: 'Magic Button',
};

export const Pressed = DefaultTemplate.bind({});

Pressed.args = {
  variant: MagicBtnVariant.Pressed,
  label: 'Magic Button',
  description: 'Magic Button Pressed',
  icon: faWandMagicSparkles,
};

export const Disabled = DefaultTemplate.bind({});

Disabled.args = {
  variant: MagicBtnVariant.Disabled,
  label: 'Magic Button',
};

export default componentMeta;
