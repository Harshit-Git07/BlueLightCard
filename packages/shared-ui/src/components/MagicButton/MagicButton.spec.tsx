import { PlatformAdapterProvider, useMockPlatformAdapter } from 'src/adapters';
import MagicButton from './';
import { Props } from './';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

const mockPlatformAdapter = useMockPlatformAdapter();

describe('Button component', () => {
  let props: Props;

  beforeEach(() => {
    props = {};
  });

  describe('smoke test', () => {
    it('should render component without error', () => {
      render(
        <PlatformAdapterProvider adapter={mockPlatformAdapter}>
          <MagicButton {...props} />
        </PlatformAdapterProvider>,
      );

      const button = screen.getByRole('button');

      expect(button).toBeTruthy();
    });
  });

  describe('snapshot Test', () => {
    it('renders a button with animated border', () => {
      const component = renderer.create(
        <PlatformAdapterProvider adapter={mockPlatformAdapter}>
          <MagicButton variant="secondary" animate>
            Button
          </MagicButton>
          ,
        </PlatformAdapterProvider>,
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('renders a button that is diabled', () => {
      const component = renderer.create(
        <PlatformAdapterProvider adapter={mockPlatformAdapter}>
          <MagicButton variant="primary" disabled>
            Button
          </MagicButton>
          ,
        </PlatformAdapterProvider>,
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
