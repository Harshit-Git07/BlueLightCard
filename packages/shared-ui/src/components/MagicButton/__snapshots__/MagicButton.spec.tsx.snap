// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MagicButton component snapshot test for component variants should render component on disabled variant 1`] = `
<button
  className=" z-10 relative w-fit rounded-full overflow-hidden p-1 h-[3.75rem] disabled"
  disabled={true}
  onClick={[Function]}
>
  <div
    className="relative py-2 px-4 rounded-full z-10 h-full w-full flex justify-center items-center text-magicButton-disabled-label-colour dark:text-magicButton-disabled-label-colour-dark bg-magicButton-disabled-bg-colour dark:bg-magicButton-disabled-bg-colour-dark opacity-30"
  >
    <span
      className="text-magicButton-default-label-colour dark:text-magicButton-default-label-colour-dark font-magicButton-label-font font-magicButton-label-font-weight text-magicButton-label-font tracking-magicButton-label-font leading-magicButton-label-font"
    >
      Button
    </span>
  </div>
</button>
`;

exports[`MagicButton component snapshot test for component variants should render component on pressed variant with icon 1`] = `
<button
  className=" z-10 relative w-fit rounded-full overflow-hidden p-1 h-[3.75rem] cursor-pointer"
  disabled={false}
  onClick={[Function]}
>
  <div
    className="absolute z-0 aspect-square min-w-[125%] min-h-[125%] top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 flex"
  >
    <div
      className="bg-gradient-to-r from-5% to-95% from-colour-secondary-gradient-bright-fixed dark:from-colour-secondary-gradient-bright-fixed via-colour-secondary-gradient-centre-fixed dark:via-colour-secondary-gradient-centre-fixed to-colour-secondary-gradient-dim-fixed dark:to-colour-secondary-gradient-dim-fixed z-0 h-full w-full animate-magicButtonGradient"
      style={
        {
          "backgroundSize": "400% 400%",
        }
      }
    />
  </div>
  <div
    className="relative py-2 px-4 rounded-full z-10 h-full w-full flex justify-center items-center bg-colour-surface dark:bg-colour-surface-dark"
  >
    <div
      className="flex-col w-full text-nowrap whitespace-nowrap flex-nowrap justify-center items-center"
    >
      <div
        className="text-magicButton-pressed-label-colour dark:text-magicButton-pressed-label-colour-dark font-magicButton-label-font font-magicButton-label-font-weight text-magicButton-label-font tracking-magicButton-label-font leading-magicButton-label-font text-center flex justify-center gap-2 items-center"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-wand-magic-sparkles "
          data-icon="wand-magic-sparkles"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={{}}
          viewBox="0 0 576 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M234.7 42.7L197 56.8c-3 1.1-5 4-5 7.2s2 6.1 5 7.2l37.7 14.1L248.8 123c1.1 3 4 5 7.2 5s6.1-2 7.2-5l14.1-37.7L315 71.2c3-1.1 5-4 5-7.2s-2-6.1-5-7.2L277.3 42.7 263.2 5c-1.1-3-4-5-7.2-5s-6.1 2-7.2 5L234.7 42.7zM46.1 395.4c-18.7 18.7-18.7 49.1 0 67.9l34.6 34.6c18.7 18.7 49.1 18.7 67.9 0L529.9 116.5c18.7-18.7 18.7-49.1 0-67.9L495.3 14.1c-18.7-18.7-49.1-18.7-67.9 0L46.1 395.4zM484.6 82.6l-105 105-23.3-23.3 105-105 23.3 23.3zM7.5 117.2C3 118.9 0 123.2 0 128s3 9.1 7.5 10.8L64 160l21.2 56.5c1.7 4.5 6 7.5 10.8 7.5s9.1-3 10.8-7.5L128 160l56.5-21.2c4.5-1.7 7.5-6 7.5-10.8s-3-9.1-7.5-10.8L128 96 106.8 39.5C105.1 35 100.8 32 96 32s-9.1 3-10.8 7.5L64 96 7.5 117.2zm352 256c-4.5 1.7-7.5 6-7.5 10.8s3 9.1 7.5 10.8L416 416l21.2 56.5c1.7 4.5 6 7.5 10.8 7.5s9.1-3 10.8-7.5L480 416l56.5-21.2c4.5-1.7 7.5-6 7.5-10.8s-3-9.1-7.5-10.8L480 352l-21.2-56.5c-1.7-4.5-6-7.5-10.8-7.5s-9.1 3-10.8 7.5L416 352l-56.5 21.2z"
            fill="currentColor"
            style={{}}
          />
        </svg>
        Button
      </div>
      <div
        className="text-magicButton-pressed-description-colour dark:text-magicButton-pressed-description-colour-dark font-magicButton-description-font font-magicButton-description-font-weight text-magicButton-description-font tracking-magicButton-description-font leading-magicButton-description-font"
      >
        Button description
      </div>
    </div>
  </div>
</button>
`;

exports[`MagicButton component snapshot test for component variants should render component on pressed variant with no icon 1`] = `
<button
  className=" z-10 relative w-fit rounded-full overflow-hidden p-1 h-[3.75rem] cursor-pointer"
  disabled={false}
  onClick={[Function]}
>
  <div
    className="absolute z-0 aspect-square min-w-[125%] min-h-[125%] top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 flex"
  >
    <div
      className="bg-gradient-to-r from-5% to-95% from-colour-secondary-gradient-bright-fixed dark:from-colour-secondary-gradient-bright-fixed via-colour-secondary-gradient-centre-fixed dark:via-colour-secondary-gradient-centre-fixed to-colour-secondary-gradient-dim-fixed dark:to-colour-secondary-gradient-dim-fixed z-0 h-full w-full animate-magicButtonGradient"
      style={
        {
          "backgroundSize": "400% 400%",
        }
      }
    />
  </div>
  <div
    className="relative py-2 px-4 rounded-full z-10 h-full w-full flex justify-center items-center bg-colour-surface dark:bg-colour-surface-dark"
  >
    <div
      className="flex-col w-full text-nowrap whitespace-nowrap flex-nowrap justify-center items-center"
    >
      <div
        className="text-magicButton-pressed-label-colour dark:text-magicButton-pressed-label-colour-dark font-magicButton-label-font font-magicButton-label-font-weight text-magicButton-label-font tracking-magicButton-label-font leading-magicButton-label-font text-center flex justify-center gap-2 items-center"
      >
        Button
      </div>
      <div
        className="text-magicButton-pressed-description-colour dark:text-magicButton-pressed-description-colour-dark font-magicButton-description-font font-magicButton-description-font-weight text-magicButton-description-font tracking-magicButton-description-font leading-magicButton-description-font"
      >
        Button description
      </div>
    </div>
  </div>
</button>
`;
